File: 00037. Sudoku Solver/solution.go
/**
 * Problem: 37. Sudoku Solver
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 1 ms (Beats 100%)
 */

// Define constants for Sudoku puzzle dimensions and Dancing Links X constraints
const (
	N           = 9
	Rows        = 9
	Cols        = 9
	Nums        = 9
	BoardSize   = Rows * Cols
	Options     = BoardSize * N
	Constraints = BoardSize * 4
)

// Column represents a constraint column in the Dancing Links X matrix
type Column struct {
	Node
	name int
	size int
}

// Node represents a matrix element in the Dancing Links X structure
type Node struct {
	up, down, left, right *Node
	col                   *Column
	rowID                 int
}

// DLX implements the Dancing Links X algorithm for exact cover problems
type DLX struct {
	head    *Node
	columns []*Column
	answer  []*Node
}

// newDLX creates a new Dancing Links X solver with specified number of columns
func newDLX(numCols int) *DLX {
	// Create header node for circular doubly-linked list
	header := &Node{}
	header.left = header
	header.right = header

	// Initialize DLX structure
	dlx := &DLX{
		head:    header,
		columns: make([]*Column, numCols),
		answer:  []*Node{},
	}

	// Build column headers in circular list
	lastCol := header

	for i := 0; i < numCols; i++ {
		// Create new column with circular vertical links
		col := &Column{name: i}
		col.up = &col.Node
		col.down = &col.Node
		col.left = lastCol
		col.right = header
		lastCol.right = &col.Node
		header.left = &col.Node
		col.col = col
		dlx.columns[i] = col
		lastCol = &col.Node
	}

	return dlx
}

// addRow adds a new row to the matrix with nodes in specified columns
func (dlx *DLX) addRow(rowID int, columns []int) {
	var firstNode *Node

	for _, colIdx := range columns {
		// Get column and create new node
		col := dlx.columns[colIdx]
		node := &Node{col: col, rowID: rowID}

		if firstNode == nil {
			// First node becomes circular list head
			firstNode = node
			node.left = node
			node.right = node
		} else {
			// Insert node into horizontal circular list
			node.left = firstNode.left	
			node.right = firstNode
			firstNode.left.right = node
			firstNode.left = node
		}

		// Insert node into vertical column list
		node.up = col.up
		node.down = &col.Node
		col.up.down = node
		col.up = node
		col.size++
	}
}

// cover removes column and all intersecting rows from the matrix
func (dlx *DLX) cover(col *Column) {
	// Remove column header from horizontal list
	col.Node.right.left = col.Node.left
	col.Node.left.right = col.Node.right

	// Remove all rows that intersect this column
	for rowNode := col.down; rowNode != &col.Node; rowNode = rowNode.down {
		for colNode := rowNode.right; colNode != rowNode; colNode = colNode.right {
			colNode.down.up = colNode.up
			colNode.up.down = colNode.down
			colNode.col.size--
		}
	}
}

// uncover restores column and all intersecting rows to the matrix
func (dlx *DLX) uncover(col *Column) {
	// Restore all rows that intersect this column
	for rowNode := col.up; rowNode != &col.Node; rowNode = rowNode.up {
		for colNode := rowNode.left; colNode != rowNode; colNode = colNode.left {
			colNode.col.size++
			colNode.down.up = colNode
			colNode.up.down = colNode
		}
	}

	// Restore column header to horizontal list
	col.Node.right.left = &col.Node
	col.Node.left.right = &col.Node
}

// search performs recursive backtracking to find exact cover solution
func (dlx *DLX) search(depth int) bool {
	// Base case: all columns covered, solution found
	if dlx.head.right == dlx.head {
		return true
	}

	// Choose column with minimum size (branching heuristic)
	minCol := dlx.head.right.col

	for colNode := dlx.head.right; colNode != dlx.head; colNode = colNode.right {
		if colNode.col.size < minCol.size {
			minCol = colNode.col
		}
	}

	// Cover chosen column
	dlx.cover(minCol)

	// Try each row in the chosen column
	for rowNode := minCol.down; rowNode != &minCol.Node; rowNode = rowNode.down {
		// Add row to partial solution
		dlx.answer = append(dlx.answer, rowNode)

		// Cover all other columns in this row
		for colNode := rowNode.right; colNode != rowNode; colNode = colNode.right {
			dlx.cover(colNode.col)
		}

		// Recursively search for solution
		if dlx.search(depth + 1) {
			return true
		}

		// Backtrack: uncover columns and remove row from solution
		for colNode := rowNode.left; colNode != rowNode; colNode = colNode.left {
			dlx.uncover(colNode.col)
		}

		dlx.answer = dlx.answer[:len(dlx.answer)-1]
	}

	// Uncover column before backtracking
	dlx.uncover(minCol)

	return false
}

// sudokuIndex converts Sudoku coordinates to unique row identifier
func sudokuIndex(row, col, num int) int {
	return row*81 + col*9 + num
}

// solveSudoku solves the Sudoku puzzle using Dancing Links X algorithm
func solveSudoku(board [][]byte) {
	// Create DLX solver with 4 types of constraints
	dlx := newDLX(Constraints)

	// Add rows for each possible cell-number combination
	for row := 0; row < 9; row++ {
		for col := 0; col < 9; col++ {
			for num := 0; num < 9; num++ {
				// Skip if cell is filled and doesn't match current number
				if board[row][col] != '.' && board[row][col] != byte(num+'1') {
					continue
				}
				
				// Create unique row ID and add constraints
				rowID := sudokuIndex(row, col, num)
				dlx.addRow(rowID, []int{
					row*9 + col,                    // Cell constraint
					81 + row*9 + num,               // Row constraint
					162 + col*9 + num,              // Column constraint
					243 + ((row/3)*3+(col/3))*9 + num, // Box constraint
				})
			}
		}
	}

	// Find solution using DLX algorithm
	dlx.search(0)

	// Fill board with solution values
	for _, solutionNode := range dlx.answer {
		rowID := solutionNode.rowID
		row := rowID / 81
		col := (rowID % 81) / 9
		num := (rowID % 9)
		board[row][col] = byte(num + '1')
	}
}

File: 00037. Sudoku Solver/solution.js
/**
 * Problem: 37. Sudoku Solver
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 7 ms (Beats 100%)
 */

/**
 * Solves a Sudoku puzzle using backtracking and bitmasking
 *
 *  @param {Array} board - 2D array representing the Sudoku board
 *
 *  @returns {void} - The board is modified in place to contain the solution
 */
const solveSudoku = (board) => {
  // Count empty cells to track remaining work
  let emptyCellCount = 0

  // Bitmasks to track used numbers in each row, column, and 3x3 box
  const rowMasks = new Array(9).fill(0),
    colMasks = new Array(9).fill(0),
    boxMasks = new Array(9).fill(0)

  // Initialize bitmasks by scanning the board
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      if (board[r][c] !== '.') {
        const boxIndex = 3 * Math.floor(r / 3) + Math.floor(c / 3)

        on(rowMasks, r, parseInt(board[r][c] - 1))
        on(colMasks, c, parseInt(board[r][c] - 1))
        on(boxMasks, boxIndex, parseInt(board[r][c] - 1))
      } else {
        emptyCellCount++
      }
    }
  }

  // Recursive backtracking function
  const backtrack = (emptyCellCount) => {
    // Base case: all cells filled
    if (emptyCellCount === 0) return true

    // Find the empty cell with minimum possible values
    const [r, c] = findMin(board, rowMasks, colMasks, boxMasks)

    const boxIndex = 3 * Math.floor(r / 3) + Math.floor(c / 3)

    // Store original bitmask values for backtracking
    const originalRowMask = rowMasks[r],
      originalColMask = colMasks[c],
      originalBoxMask = boxMasks[boxIndex]

    // Get intersection of constraints
    const constraintMask = rowMasks[r] | colMasks[c] | boxMasks[boxIndex]

    // Try each possible number (1-9)
    for (let i = 0; i < 9; i++) {
      if (has(constraintMask, i)) {
        // Place the number
        board[r][c] = (i + 1).toString()

        // Update bitmasks
        on(rowMasks, r, i)
        on(colMasks, c, i)
        on(boxMasks, boxIndex, i)

        // Recurse with one less empty cell
        if (backtrack(emptyCellCount - 1)) return true

        // Backtrack: remove the number and restore bitmasks
        board[r][c] = '.'
        rowMasks[r] = originalRowMask
        colMasks[c] = originalColMask
        boxMasks[boxIndex] = originalBoxMask
      }
    }

    return false
  }

  // Start the backtracking process
  backtrack(emptyCellCount)
}
/**
 * Sets bit at position v in array element at index i
 *
 *  @param {Array} arr - The array of bitmasks
 *  @param {number} i - The index in the array to modify
 *  @param {number} v - The bit position to set (0-8 for
 *
 * @returns {void}
 */
const on = (arr, i, v) => (arr[i] |= 1 << v)

/**
 * Checks if bit at position is NOT set (number is available)
 *
 * @param {number} val - The bitmask value to check
 * @param {number} bit - The bit position to check (0-8 for Sudoku)
 *
 * @returns {boolean} - True if the bit is NOT set, meaning the number is available
 *
 */
const has = (val, bit) => (val & (1 << bit)) === 0

/**
 * Finds empty cell with minimum possible values (MRV heuristic)
 *
 * @param {Array} board - The Sudoku board
 * @param {Array} rows - Bitmasks for each row
 * @param {Array} cols - Bitmasks for each column
 * @param {Array} boxes - Bitmasks for each 3x3 box
 *
 * @returns {Array} - Coordinates of the cell with minimum possibilities [row, col]
 */
const findMin = (board, rows, cols, boxes) => {
  // Initialize variables to track the empty cell with minimum possible values
  let minRow = -1,
    minCol = -1,
    minPossibleCount = Infinity

  // Search through all cells to find the one with fewest possibilities
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      if (board[r][c] === '.') {
        // Calculate box index for current cell
        const boxIndex = 3 * Math.floor(r / 3) + Math.floor(c / 3)
        let possibleCount = 0

        // Get combined constraints from row, column, and box
        const combinedConstraints = rows[r] | cols[c] | boxes[boxIndex]

        // Count how many numbers are still possible for this cell
        for (let i = 0; i < 9; i++)
          if (has(combinedConstraints, i)) possibleCount++

        // Update minimum if this cell has fewer possibilities
        if (possibleCount < minPossibleCount) {
          minPossibleCount = possibleCount
          minRow = r
          minCol = c
        }
      }
    }
  }

  return [minRow, minCol]
}
